src/reconcile/rec_visibility.c

```diff
@@ -414,12 +414,47 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
 
             /* Find the update this tombstone applies to. */
             if (!__wt_txn_upd_visible_all(session, upd)) {
-                while (upd->next != NULL && upd->next->txnid == WT_TXN_ABORTED)
-                    upd = upd->next;
+                /*
+                 * Loop until a valid update from a different transaction is found in the update
+                 * list.
+                 */
+                while (upd->next != NULL) {
+                    if (upd->next->txnid == WT_TXN_ABORTED)
+                        upd = upd->next;
+                    else if (tombstone->start_ts == upd->next->start_ts &&
+                      tombstone->durable_ts == upd->next->durable_ts &&
+                      tombstone->txnid == upd->next->txnid) {
+                        upd = upd->next;
+                        same_txn_valid_upd = upd;
+                    } else
+                        break;
+                }
+
                 WT_ASSERT(session, upd->next == NULL || upd->next->txnid != WT_TXN_ABORTED);
                 if (upd->next == NULL)
                     last_upd = upd;
                 upd_select->upd = upd = upd->next;
+
+                /*
+                 * If there is no on-disk update and any valid update from a different transaction
+                 * is not found in the update list, write the same transaction update itself to disk
+                 * to avoid blocking the eviction.
+                 */
+                if (vpack == NULL && upd == NULL)
+                    upd_select->upd = upd = same_txn_valid_upd;
+                else if (upd != NULL && upd->type == WT_UPDATE_TOMBSTONE)
+                    /*
+                     * The selected update from a different transaction is also a tombstone, use the
+                     * update from same transaction as the selected update.
+                     */
+                    upd_select->upd = upd = same_txn_valid_upd;
+                else if (same_txn_valid_upd != NULL && vpack != NULL &&
+                  WT_TIME_WINDOW_HAS_STOP(&vpack->tw))
```

Comment:
Why we need to write same_txn_valid_upd to disk in this case?

Reply:
If we don't write the same transaction to disk in this scenario, these updates must be restored either with update restore to let the prepare transaction to either commit or rollback later. As the on-disk version already has a tombstone, so we cannot apply the prepared tombstone to it. If we don't write these updates to disk, it blocks the eviction.

Reply:
Thanks I get it now.